### PR TITLE
tsdb store refactoring

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
It includes two small changes:
1. We made tsdb store instance public to only allow creating a single instance of it. I have refactored the code not to make it public.
2. Build a tsdb from active wal and upload it to the object storage during ingester shutdown. This is to avoid losing the index if an ingester is being shut down forever.

### CODE REVIEW

File: pkg/storage/stores/tsdb/store.go [hunk 1]
```diff
@@ -17,37 +19,69 @@ import (
...
 type store struct {
-	indexWriter IndexWriter
-	indexStore  series.IndexStore
+	indexShipper indexshipper.IndexShipper
+	indexWriter  IndexWriter
+	indexStore   series.IndexStore
+	stopOnce     sync.Once
 }
 
+type newStoreFactoryFunc func(
```
Comment (jeschkies) [inline pkg/storage/stores/tsdb/store.go:29]: Is this used anywhere else than for `NewStore`?

File: pkg/storage/stores/tsdb/store.go [hunk 2]
```diff
@@ -17,37 +19,69 @@ import (
...
+			err := storeInstance.init(indexShipperCfg, objectClient, limits, tableRanges, reg)
+			if err != nil {
+				return nil, nil, nil, err
+			}
 		}
-	}
 
-	return NewChunkWriter(f, p, storeInstance.indexWriter), storeInstance.indexStore, nil
-}
+		return NewChunkWriter(f, p, storeInstance.indexWriter), storeInstance.indexStore, storeInstance.Stop, nil
```
Comment (jeschkies) [inline pkg/storage/stores/tsdb/store.go:76]: Why not return the `storeInstance`?

File: pkg/storage/stores/tsdb/store.go (hunk 1 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/tsdb/store.go:29]: No, its not. It is just for defining function signature.

File: pkg/storage/stores/tsdb/store.go (hunk 2 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/tsdb/store.go:76]: For which type? We can't do it for writer atleast as explained here https://github.com/grafana/loki/blob/d6f0455b656d9ad2c99aabcda5ec2dbbbd6135a8/pkg/storage/stores/tsdb/store.go#L45

Comment (owen-d) [reply pkg/storage/stores/tsdb/store.go:76]: @jeschkies Yeah, we should refactor to make this more intuitive/less ugly.

**Pull request merged**